File: pet-adoption/backend/petPalFinder/src/main/java/com/petAdoption/petPalFinder/services/EmailVerificationService.java
package com.petAdoption.petPalFinder.services;

import com.petAdoption.petPalFinder.dto.OtpVerificaion;
import com.petAdoption.petPalFinder.dto.StatusMessage;
import com.petAdoption.petPalFinder.dto.StatusUpdateDto;
import com.petAdoption.petPalFinder.models.EmailVerification;

public interface EmailVerificationService {
    public StatusMessage sendSimpleMail(EmailVerification email);
    public StatusMessage verifyOtp(OtpVerificaion otpVerificaion);
    public void sendStatusMail(String email, StatusUpdateDto status, String role);
}

File: pet-adoption/frontend/pet-pal-finder/src/app/service/email-verification.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import {
  EmailVerificationRequest,
  Otp,
  StatusMessage,
  User,
} from '../models/models';

@Injectable({
  providedIn: 'root',
})
export class EmailVerificationService {
  private userSubject = new BehaviorSubject<User>({
    email: '',
    role: '',
    password: '',
  });
  sharedUser$ = this.userSubject.asObservable();
  private emailSubject = new BehaviorSubject<string>('');
  sharedEmail$ = this.emailSubject.asObservable();
  private roleSubject = new BehaviorSubject<string>('');
  sharedRole$ = this.roleSubject.asObservable();

  private baseUrl: string = environment.baseUrl;

  constructor(private http: HttpClient) {}

  verifyEmail(
    emailVerificationRequest: EmailVerificationRequest
  ): Observable<StatusMessage> {
    this.emailSubject.next(emailVerificationRequest.email);
    const status: Observable<StatusMessage> = this.http.post<any>(
      `${this.baseUrl}email-verification`,
      emailVerificationRequest
    );
    return status;
  }

  verifyOtp(otp: Otp): Observable<StatusMessage> {
    const status: Observable<StatusMessage> = this.http.post<any>(
      `${this.baseUrl}email-verification/verify`,
      otp
    );
    return status;
  }
}
